File: macos/Tests/NoctusCoreTests/CommandRegistryTests.swift
import Testing
@testable import NoctusCore

@Suite struct CommandRegistryTests {

    func makeCommand(_ id: String, canExec: Bool = true) -> CommandDefinition {
        CommandDefinition(id: id, name: id, canExecute: { canExec }, execute: {})
    }

    @Test func registerAndGetById() {
        let reg = CommandRegistry()
        reg.register(makeCommand("test"))
        #expect(reg.getById("test") != nil)
    }

    @Test func getByIdUnknownReturnsNil() {
        let reg = CommandRegistry()
        #expect(reg.getById("nope") == nil)
    }

    @Test func getAllReturnsAll() {
        let reg = CommandRegistry()
        reg.register(makeCommand("a"))
        reg.register(makeCommand("b"))
        #expect(reg.getAll().count == 2)
    }

    @Test func executeCallsAction() {
        let reg = CommandRegistry()
        var called = false
        reg.register(CommandDefinition(id: "t", name: "t", execute: { called = true }))
        reg.execute("t")
        #expect(called == true)
    }

    @Test func executeWhenCannotDoesNothing() {
        let reg = CommandRegistry()
        var called = false
        reg.register(CommandDefinition(id: "t", name: "t", canExecute: { false }, execute: { called = true }))
        reg.execute("t")
        #expect(called == false)
    }

    @Test func canExecuteReturnsCorrectly() {
        let reg = CommandRegistry()
        reg.register(makeCommand("yes", canExec: true))
        reg.register(makeCommand("no", canExec: false))
        #expect(reg.canExecute("yes") == true)
        #expect(reg.canExecute("no") == false)
        #expect(reg.canExecute("nope") == false)
    }
}

File: tests/Explorer.Core.Tests/Services/CommandRegistryTests.cs
using NoctusExplorer.Core.Models;
using NoctusExplorer.Core.Services;
using FluentAssertions;

namespace NoctusExplorer.Core.Tests.Services;

public class CommandRegistryTests
{
    private static CommandDefinition MakeCommand(string id, bool canExecute = true)
    {
        bool executed = false;
        return new CommandDefinition
        {
            Id = id,
            Name = id.Replace('.', ' '),
            CanExecute = () => canExecute,
            Execute = () => executed = true
        };
    }

    [Fact]
    public void Register_AddsCommand()
    {
        var reg = new CommandRegistry();
        reg.Register(MakeCommand("test.cmd"));
        reg.GetById("test.cmd").Should().NotBeNull();
    }

    [Fact]
    public void Register_DuplicateId_Throws()
    {
        var reg = new CommandRegistry();
        reg.Register(MakeCommand("test.cmd"));
        var act = () => reg.Register(MakeCommand("test.cmd"));
        act.Should().Throw<ArgumentException>();
    }

    [Fact]
    public void GetById_UnknownId_ReturnsNull()
    {
        var reg = new CommandRegistry();
        reg.GetById("nope").Should().BeNull();
    }

    [Fact]
    public void GetAll_ReturnsAllRegistered()
    {
        var reg = new CommandRegistry();
        reg.Register(MakeCommand("a"));
        reg.Register(MakeCommand("b"));
        reg.GetAll().Should().HaveCount(2);
    }

    [Fact]
    public void Execute_CallsCommandAction()
    {
        var reg = new CommandRegistry();
        bool wasCalled = false;
        reg.Register(new CommandDefinition
        {
            Id = "test",
            Name = "Test",
            CanExecute = () => true,
            Execute = () => wasCalled = true
        });
        reg.Execute("test");
        wasCalled.Should().BeTrue();
    }

    [Fact]
    public void Execute_WhenCannotExecute_DoesNothing()
    {
        var reg = new CommandRegistry();
        bool wasCalled = false;
        reg.Register(new CommandDefinition
        {
            Id = "test",
            Name = "Test",
            CanExecute = () => false,
            Execute = () => wasCalled = true
        });
        reg.Execute("test");
        wasCalled.Should().BeFalse();
    }

    [Fact]
    public void Execute_UnknownId_DoesNothing()
    {
        var reg = new CommandRegistry();
        var act = () => reg.Execute("nope");
        act.Should().NotThrow();
    }

    [Fact]
    public void CanExecute_ReturnsDelegateResult()
    {
        var reg = new CommandRegistry();
        reg.Register(MakeCommand("yes", canExecute: true));
        reg.Register(MakeCommand("no", canExecute: false));
        reg.CanExecute("yes").Should().BeTrue();
        reg.CanExecute("no").Should().BeFalse();
    }

    [Fact]
    public void CanExecute_UnknownId_ReturnsFalse()
    {
        var reg = new CommandRegistry();
        reg.CanExecute("nope").Should().BeFalse();
    }
}
